modules/new_serverpod_auth/serverpod_auth_profile/serverpod_auth_profile_server/lib/src/business/user_profiles.dart

```diff
@@ -0,0 +1,404 @@
+import 'dart:isolate';
+import 'dart:typed_data';
+
+import 'package:http/http.dart' as http;
+import 'package:image/image.dart';
+import 'package:serverpod/serverpod.dart';
+import 'package:serverpod_auth_profile_server/serverpod_auth_profile_server.dart';
+import 'package:serverpod_auth_profile_server/src/generated/protocol.dart';
+import 'package:serverpod_auth_profile_server/src/util/user_profile_extension.dart';
+
+/// Business logic for handling user profiles
+abstract final class UserProfiles {
+  /// Creates a new user and stores it in the database.
+  static Future<UserProfileModel> createUserProfile(
+    final Session session,
+    final UuidValue authUserId,
+    UserProfileData userProfile,
+  ) async {
+    final onBeforeUserProfileCreated =
+        UserProfileConfig.current.onBeforeUserProfileCreated;
+    if (onBeforeUserProfileCreated != null) {
+      userProfile = await onBeforeUserProfileCreated(
+        session,
+        authUserId,
+        userProfile,
+      );
+    }
+
+    userProfile = userProfile.copyWith(
+      email: userProfile.email?.toLowerCase().trim(),
+    );
+
+    final createdProfile = await UserProfile.db.insertRow(
+      session,
+      UserProfile(
+        authUserId: authUserId,
+        userName: userProfile.userName,
+        fullName: userProfile.fullName,
+        email: userProfile.email,
+      ),
+    );
+
+    final createdProfileModel = createdProfile.toModel();
+
+    await UserProfileConfig.current.onAfterUserProfileCreated?.call(
```

Comment:
@SandPod What do you think we should do here:

- State that the `onAfterUserProfileCreated` is for information only and any exceptions thrown ignored (though still await-ed, or is even that too much in that case)?
- Use a transaction for the above insert, and roll back in case the `onAfterUserProfileCreated` fails?

Like in the "before handler" discussion, we might just be better off removing it and letting the application code run any post-setup code itself.
That should be much clearer on the call side, and we don't have to impose either of the 2 rules (which are basically opposites, so one kind of implementation will not be helped by them, so you can not bend them to work in either case).

Reply:
As dicussed, keeping the hooks for now, but giving it the transaction in case they would need to look at the "in progress" database state.

Since we don't have dependency injection yet and could not exchange a `UserProfiles` implementation for an adapted one, it would not be realistic to change all callers into this module (which might be many layers above), so there is currently a good case to be made for these hooks, even if they result in a less "predictable" module implementation).